### PR TITLE
[ENT-227] Restore Django 1.10 support

### PR DESCRIPTION
**Pull request by snowcrshd**
**Description:** Add back support for Django 1.10, which was dropped on #44

**Sandbox link**: https://pr15273.sandbox.opencraft.hosting

**JIRA:** https://openedx.atlassian.net/browse/ENT-227

**Dependencies:** 

**Merge deadline:** 

**Installation instructions:**

**Testing instructions:**

Sandbox

1. Log into the edx-platform sandbox as one of default users.
2. Navigate around
3. Verify everything works as normal

Unit tests

Simply run `tox`.

**Merge checklist:**

- [x] ~Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.~ Had to manually set the new requirements, because new pylint,pycodestyle,pydocstyle versions give **way** to many errors to fix in this PR.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] ~Documentation updated (not only docstrings)~ N/A
- [x] Commits are (reasonably) squashed
- [ ] ~Translations are updated~ N/A
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** Pylint is unpinned ATM. This caused an **incredible** amount of errors when running the quality checks, due to the updated requirements files. I still have to fix them before this PR can be merged.

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @snowcrshd! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=119&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

**Comment by snowcrshd** [comment]
Phew, fixed a lot of quality errors.

There are a lot more still, though 😭 

I'm getting a bunch of complaints from `pydocstyle` and couldn't figure out the reason of
`isort`'s complaints.

@e-kolpakov I should have this ready soon.
It was a bit more work than what I had anticipated.

File: integrated_channels/sap_success_factors/transmitters/courses.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def transmit_block(self, serialized_payload):
 
         return status_code, body
 
-    def transmit(self, course_exporter):
+    def transmit(self, course_exporter):  # pylint: disable=arguments-differ
```
Comment (e-kolpakov) [inline integrated_channels/sap_success_factors/transmitters/courses.py]: @snowcrshd `arguments-differ` => same signature as base method + `unused-args`

File: tests/test_learner_data.py [hunk 2]
```diff
@@ -60,7 +60,8 @@ def setUp(self):
 
     def test_collect_learner_data_no_enrollments(self):
         learner_data = list(self.exporter.collect_learner_data())
-        assert len(learner_data) == 0
+        learner_data_count = len(learner_data)
+        assert learner_data_count == 0
```
Comment (e-kolpakov) [inline tests/test_learner_data.py]: @snowcrshd Had pylint complained about this? What was the message? (this is really weird, probably false positive)

File: tests/test_management.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def setUp(self):
 
     def test_enterprise_customer_not_found(self):
         faker = FakerFactory.create()
-        invalid_customer_id = faker.uuid4()
+        invalid_customer_id = faker.uuid4()  # pylint: disable=no-member
```
Comment (e-kolpakov) [inline tests/test_management.py]: @snowcrshd Almost definitely false positive (and it didn't complain about this before). New version of pylint being more rigorous than he old one? Or maybe a plugin is missing?

File: tests/admin/test_view.py [hunk 4]
```diff
@@ -1138,7 +1190,7 @@ def test_delete_not_linked(self):
     def test_delete_linked(self):
         self._login()
 
-        email = FAKER.email()
+        email = FAKER.email()  # pylint: disable=no-member
```
Comment (e-kolpakov) [inline tests/admin/test_view.py]: @snowcrshd If there are really no better way  to make pylint understand faker (i.e. plugin), I'd suggest to improve it a little bit:

* Creating a dedicated pylintrc file in `tests` folder - copy from top-level pylintrc (better, reference it, but I haven't found a way to do that)
* Put additional suppressions into it:
    * no-member
    * that rule that requires putting `len(array)` into variable first
* Change pylint run commands in travis.yml, tox.ini and Makefile to use it - pylint does not consider pylintrc files in nested folders as overrides, so you'd have to use pylint --rcfile=tests/pylintrc)

File: tests/test_management.py [hunk 5]
```diff
@@ -22,6 +20,8 @@
 from django.utils import timezone
 
 from enterprise import lms_api
+from integrated_channels.integrated_channel.learner_data import BaseLearnerExporter
```
Comment (e-kolpakov) [inline tests/test_management.py]: @snowcrshd Hint: `isort` with `--check-only` checks that imports are sorted as it see fit (there's config somewhere, maybe in tox.ini). Without `--check-only` it'll actually sort imports itself.

But I'm a bit puzzled why this change is needed in the first place - isort didn't complain here previously (AFAIK).

File: tox.ini [hunk 6]
```diff
@@ -57,7 +60,7 @@ commands =
     touch tests/__init__.py
     pylint enterprise
     pylint integrated_channels
-    pylint tests
+    pylint tests --load-plugins=pylint.extensions.redefined_variable_type
```
Comment (e-kolpakov) [inline tox.ini]: @snowcrshd this could go into tests/pylintrc (if I recall right that you can specify plugins via pylintrc)

File: test_utils/fake_catalog_api.py [hunk 7]
```diff
@@ -481,8 +481,10 @@ def get_catalog_courses(catalog_id):
     Arguments:
         catalog_id (int): Catalog ID
 
-    Returns:
+    Returns
+    -------
```
Comment (e-kolpakov) [inline test_utils/fake_catalog_api.py]: @snowcrshd Please run `make docs` and make sure these underlining does not break documentation layout and does not introduce "Return" records into navigation menu (left sidebar).

File: enterprise/api/v1/views.py [hunk 8]
```diff
@@ -229,7 +229,7 @@ def retrieve(self, request, pk=None):  # pylint: disable=invalid-name
             request (HttpRequest): Current request
             pk (int): Course catalog identifier
 
-        Returns:
+        Returns
```
Comment (e-kolpakov) [inline enterprise/api/v1/views.py]: @snowcrshd Inconsistentsy: in this file (and some others) "Returns" are not underlined, in some other files it is. I.e. here it is just

    Returns

while in those other files it is

    Returns
    -----------

**Comment by snowcrshd** [comment]
@e-kolpakov thank you for the suggestions!

I decided to move the quality changes to another branch and just upgrade
the necessary requirements.

If in the future it is of interest to migrate to newer versions of pylint, pydocstyle, etc, I'll still have the changes around.

I'm a bit stuck at running jasmine tests on my system, though.
All other checks pass, but I'm unfamiliar with Jasmine.

**Comment by e-kolpakov** [comment]
@snowcrshd Overall looks good, but:

* There're some conflicts (likely osmeone just merged another PR which also updated the version - should be trivial to resolve)
* I don't think it's a good idea to merge this PR just yet as it references your working branch from DCM - I'd released a new version of DCM first (not 100% if it happens as part of travis build - you might need to talk to edX people to release it), than used that version here.

And I haven't yet tested it with actual edx-platform - there might be more surprises. If you haven't tried installing it into edx-platform locally and using it - you might want to do that to address any issues proactively.

**Comment by snowcrshd** [comment]
@e-kolpakov I'll solve the conflicts and update the PR.

About DCM, that's what I am waiting for, actually.
I just added that additional commit for the dependencies links so we could test it.

I'll setup get edx-platform and run their tests locally, using this PR code and DCM's updated version.

**Comment by snowcrshd** [comment]
Rebased PR and updated the release version.

**Comment by snowcrshd** [comment]
Rebased on master, to update changelog and and version.

File: CHANGELOG.rst [hunk 9]
```diff
@@ -25,7 +31,7 @@ Unreleased
 
 * Update data backing and behavior of enterprise landing page
 * Fix template prioritization bug
-* Fix URL rendering in enterprise login decorator
```
Comment (e-kolpakov) [inline CHANGELOG.rst]: @snowcrshd You sure this line shouldn't be there?

Comment (snowcrshd) [inline CHANGELOG.rst]: @e-kolpakov thank you for spotting that.

fixed.

**Comment by snowcrshd** [comment]
Rebased on top of master to update CHANGELOG and version

**Comment by snowcrshd** [comment]
I keep getting failures due to the atomic transactions introduced by #109.

Could someone, please, retrigger the failed job? I don't have permission to do that.

**Comment by bradenmacdonald** [comment]
@snowcrshd Who was triggering builds for you before? I don't see an option to do that for some reason.

**Comment by snowcrshd** [comment]
@bradenmacdonald, @haikuginger retriggered one for me yesterday.

File: requirements/base.in [hunk 10]
```diff
@@ -14,5 +14,5 @@ django_extensions                       # Required to use TimeStampedModel
 django-simple-history                   # History for Django models
 django-object-actions                   # Object actions in Django admin
 edx-rest-api-client                     # For accessing the Enrollment API (and possibly other edX APIs)
-django-config-models                    # Provides ConfigurationModel abstract base class
+-e git://github.com/snowcrshd/django-config-models.git@lucas/OC-2387/django-1.10-support#egg=django-config-models
```
Comment (brittneyexline) [inline requirements/base.in]: Is there a reason we need this package override now that the django config models change has been merged?

Comment (snowcrshd) [inline requirements/base.in]: @brittneyexline that's just for testing while DCM is not published.
I'll update the commit when this is good.

**Comment by snowcrshd** [comment]
Rebased on top of a87038ae6734f5bf7ff322afc9ae8d63ada5eab0

**Comment by snowcrshd** [comment]
DCM 0.1.6 has been published.

I updated the PR to reflect this.

@e-kolpakov @brittneyexline please check the provisioned sandbox: https://pr15273.sandbox.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/info

cc @bradenmacdonald

File: setup.py [hunk 11]
```diff
@@ -49,12 +49,12 @@ def get_version(*file_paths):
     ],
     include_package_data=True,
     install_requires=[
-        "Django>=1.8.15,<1.10",
-        "djangorestframework>=3.2.3,<3.3.0",
-        "django-config-models",
+        "Django>=1.8.15,<1.11",
+        "djangorestframework>=3.2.3,<3.7.0",
+        "django-config-models==0.1.6",
```
Comment (e-kolpakov) [inline setup.py]: @snowcrshd nit: I'd suggest >=0.1.6, but this can be addressed later when (and if) next DCM version comes out.

Comment (snowcrshd) [inline setup.py]: @e-kolpakov yeah, I think we can leave it like ==0.1.6 for now.

TBH, I prefer to pin a specific version to make deploys deterministic.

**Comment by snowcrshd** [comment]
@brittneyexline anything else you think needs change in this PR?

😸

**Comment by snowcrshd** [comment]
Squashing commits so we can merge this.

**Comment by snowcrshd** [comment]
I don't have permission to merge this. Could someone do the honors?

cc @bradenmacdonald

**Comment by itsjeyd** [comment]
@snowcrshd I'll merge this PR for you.

**Comment by snowcrshd** [comment]
@itsjeyd thank you :smiley:

**Comment by itsjeyd** [comment]
@brittneyexline @douglashall For some reason, GitHub didn't create a tag for this, and the new version is not on PyPI. Any idea why?

**Comment by brittneyexline** [comment]
@itsjeyd the tags are not created automatically - you have to go to the releases tab and create a new release with a new tag pointing to the new commit on master. if you don't have permission to do it, i can take care of it, just let me know.

**Comment by itsjeyd** [comment]
@brittneyexline That's weird, I didn't have to do anything to create the [previous tag](https://github.com/edx/edx-enterprise/releases/tag/0.34.4). In any case, we have a new release now: https://github.com/edx/edx-enterprise/releases/tag/0.34.5. It's not up on PyPI yet though.

**Comment by brittneyexline** [comment]
@itsjeyd Hmm, weird. I always publish a release after bumping the version so I wasn't aware there was any other mechanism for creating tags in this repo. Once the travis build passes on master it should be on PyPI.

**Comment by douglashall** [comment]
@itsjeyd Apologies, I created the 0.34.4 tag. I should have made you aware of this.

**Comment by itsjeyd** [comment]
@douglashall No worries, and thanks for the info! The Travis [build](https://travis-ci.org/edx/edx-enterprise/builds/242444854) passed, and the release is up on PyPI: https://pypi.python.org/pypi/edx-enterprise/0.34.5.

CC @brittneyexline @snowcrshd
